Add `MultiCallSendOnly` support to Safe{Wallet} guard

duncancmt opened the pull request:
In response to https://github.com/0xProject/0x-settler/pull/253#issuecomment-3208533193

duncancmt committed:
Fix CI

duncancmt committed:
WIP: add a Safe Guard that implements a timelock for the upgrade Safe

duncancmt committed:
Cleanup

duncancmt committed:
Cleanup

duncancmt committed:
Bug! Wrong direction on timelock checking; add check that transaction has been queued

duncancmt committed:
Add `cancel` function to Timelock/Guard

duncancmt committed:
Revert "Cleanup"

This reverts commit 307bb0a37175c3e3732f88a070295d1768445ed5.

duncancmt committed:
Add lockdown mode to Safe Guard (requires unanimous signatures to unlock)

duncancmt committed:
Comment

duncancmt committed:
Comments

duncancmt committed:
Simplify guard design by storing only a single `txHash` and expiry instead of mapping from the hash to the expiry

duncancmt committed:
Typo

duncancmt committed:
Comment

duncancmt committed:
Make initial `delay` a constructor argument

duncancmt committed:
Bug! Prevent smuggling calls to `unlock` through `enqueue` => multicall => `unlock`

duncancmt committed:
Remove assumption about the relationship between `encodeTransactionData` and `getTransactionHash`

duncancmt committed:
Resolve TODO

duncancmt committed:
Comment

duncancmt committed:
Rework Safe guard so that it only requires 1 multisig transaction; this improves the UX of using the guard

It still requires 2 transactions to be submitted. There's no way around that.

duncancmt committed:
Remove unused

duncancmt committed:
Add some extra safety checks; realistically they should never trigger, but we want to wear tinfoil

duncancmt committed:
Attempt to guard against as many footguns and bizarre scenarios as possible

duncancmt committed:
Golf

duncancmt committed:
Add some indexed fields to `SafeTransactionEnqueued` event

duncancmt committed:
Explain provenance of copied code

duncancmt committed:
Golf

duncancmt committed:
Explain comment a little better (less ambiguous)

duncancmt committed:
Pedantry

duncancmt committed:
Remove redundant check

duncancmt committed:
Resolve TODO

duncancmt committed:
Add happy-path unit test for Safe Guard

duncancmt committed:
Add timelock sad path test case for Guard

duncancmt committed:
Bug! Require that a call to `cancel` be accompanied by approving the `unlock` transaction hash; this prevents a rogue owner from repeatedly calling `cancel` to brick the safe

duncancmt committed:
Fix CI

duncancmt committed:
Prevent calling the specified Guard functions outside of the Guard context

duncancmt committed:
Fix Safe owner reverse order

duncancmt committed:
Check that the Safe call is actually going through

duncancmt committed:
I give up; completely remove the ability to `DELEGATECALL` from the Safe

duncancmt committed:
Remove unnecessary duplicate locked down checks

duncancmt committed:
Cleanup

duncancmt committed:
Update copied interface in test

duncancmt committed:
Bug! Prevent re-queueing transaction (possibly clearing cancelation)

duncancmt committed:
Fail early if attempting an `Operation.DelegateCall`

duncancmt committed:
Check event

duncancmt committed:
Add canceller address to event

duncancmt committed:
Test for `DELEGATECALL` prevention

duncancmt committed:
Test for `cancel`

duncancmt committed:
Add some sad-path tests for `cancel`

duncancmt committed:
Add test for `lockDown`

duncancmt committed:
Add test for `unlock`

duncancmt committed:
Cleanup

duncancmt committed:
Add test for non-unanimous `unlock`

duncancmt committed:
Replace `notRemoved` and `notLockedDown` modifiers witih `normalOperation` modifier that combines them; apply this modifier to `enqueue` and `lockDown`

duncancmt committed:
Comment

duncancmt committed:
Check for emitted success event by Safe

duncancmt committed:
Remove noisy debug logging

duncancmt committed:
Cleanup

duncancmt committed:
Merge branch 'master' into dcmt/upgrade-timelock

duncancmt committed:
Merge branch 'master' into dcmt/upgrade-timelock

duncancmt committed:
Try to speed up CI

duncancmt committed:
Add Dedaub audit report for `SafeGuard.sol`

duncancmt committed:
Add *DRAFT* Bailsec audit report for `SafeGuard.sol`

duncancmt committed:
Address Bailsec audit findings

duncancmt committed:
`forge fmt`

duncancmt committed:
Add graffiti event to show the resignation `txHash`

duncancmt committed:
Fix unit tests

duncancmt committed:
Add some new test cases

duncancmt committed:
Add another new test case

duncancmt committed:
`forge fmt`

duncancmt committed:
Add unit test for the `resignTxHash` flow

duncancmt committed:
DRY

duncancmt committed:
Add final Bailsec audit report for `SafeGuard.sol`

duncancmt committed:
Reorder checks for clarity

duncancmt committed:
Make `safe` a constructor argument in `SafeGuard`

duncancmt committed:
`forge fmt`

duncancmt committed:
Support both Arachnid deployer and Safe singleton factory rooted Safe{Wallet} deployments

duncancmt committed:
Comment

duncancmt committed:
Add more unit test

duncancmt committed:
Merge branch 'master' into dcmt/upgrade-timelock

duncancmt committed:
Merge branch 'master' into dcmt/upgrade-timelock

duncancmt committed:
Golf

duncancmt committed:
Revert "Golf"

This reverts commit 95f94997808282cd5b6852d8cdb98de4a4a60649.

duncancmt committed:
Fix CI

duncancmt committed:
Fix CI

duncancmt committed:
Style

duncancmt committed:
Add a variation of the Safe{Wallet} Guard timelock for Safe{Wallet} version 1.4.1

duncancmt committed:
Add some `vm.expectCall`

duncancmt committed:
Avoid opaque constants in test files

duncancmt committed:
Avoid opaque constants in test files

duncancmt committed:
Add references to canonical Safe{Wallet} 1.4.1 deployment addresses

duncancmt committed:
Check Safe{Wallet} proxy code hash on deployment

duncancmt committed:
Add reference to Safe{Wallet} 1.1.1 factory, because that is also supported

duncancmt committed:
Give names to constants

duncancmt committed:
`abstract`

duncancmt committed:
Typo

duncancmt committed:
Typo

duncancmt committed:
Allow calls to `MultiCallSendOnly`, but use deep inspection to forbid `enableModule`

duncancmt committed:
Also check the value of the fallback handler after each transaction

duncancmt committed:
`forge fmt`

duncancmt committed:
Compilation errors

duncancmt committed:
Add comment in response to Immunefi Magnus PR review

duncancmt committed:
Be explicit about differences in our decoding of multicall and `MultiSendCallOnly`

duncancmt committed:
Cleanup

duncancmt committed:
Golf

duncancmt committed:
Bug! Reversed `if` condition

duncancmt committed:
Cleanup

duncancmt committed:
Bug! You have to check the selector when blacklisting a specific selector

duncancmt committed:
`forge fmt`

duncancmt committed:
Check Safe{Wallet} Guard invariants in between user-defined multicall operations

Review by V12 agent uncovered several bugs in the handling of multicall deep introspection

duncancmt committed:
Merge branch 'master' into dcmt/upgrade-timelock-1.4.0-multicall

duncancmt committed:
`forge fmt`

duncancmt committed:
Forbid calls to the Safe{Wallet} Guard's `checkAfterExecution` in the middle of a multicall

This bug also found by V12

duncancmt committed:
Correctly handle resignations on a Safe{Wallet} configured to be unanimous

This bug also found by V12

duncancmt committed:
Compilation errors

duncancmt committed:
Variable naming consistency

duncancmt committed:
`forge fmt`

duncancmt committed:
Add event

duncancmt committed:
WIP: Fixing tests

duncancmt committed:
`forge fmt`

duncancmt committed:
Fix tests

duncancmt committed:
`unchecked`

duncancmt committed:
Forbid `checkTransaction` inside multicall

duncancmt committed:
`unchecked`

duncancmt committed:
Golf

duncancmt committed:
Comment

duncancmt committed:
Golf

duncancmt committed:
DRY

duncancmt committed:
Compilation errors

duncancmt committed:
Symmetry; golf

duncancmt committed:
Simplify

duncancmt committed:
Add support for ERC-7955 toehold

duncancmt committed:
Add explanation/enumeration of Safe{Wallet} Guard failure cases

duncancmt committed:
Improve explanation of Safe{Wallet} Guard failure cases

duncancmt committed:
Improve explanation of Safe{Wallet} Guard failure cases

duncancmt committed:
Improve explanation of Safe{Wallet} Guard failure cases

duncancmt committed:
Rename error

duncancmt committed:
Reconfigure minimum owner/threshold checks

duncancmt committed:
Disallow canceling one's own resignation

duncancmt committed:
Merge branch 'master' into dcmt/upgrade-timelock-1.4.0-multicall

duncancmt committed:
Compilation error

duncancmt committed:
Fixing tests

duncancmt committed:
Fixing tests

duncancmt committed:
Pedantry

duncancmt committed:
Merge branch 'master' into dcmt/upgrade-timelock-1.4.0-multicall

duncancmt committed:
Add zkSync support to SafeGuard

duncancmt committed:
Clean up slop

duncancmt committed:
`forge fmt`

duncancmt committed:
Fix SafeGuard EraVM harness and tests

duncancmt committed:
Clean up slop

duncancmt committed:
Refine SafeGuard EraVM tests

duncancmt committed:
WIP: debugging EraVm

duncancmt committed:
Fix SafeGuard EraVM dummy hash and zk test config

duncancmt committed:
Clean up slop

duncancmt committed:
Fix SafeGuard EraVM dummy hash and zk test config

duncancmt committed:
Fix foundry-zksync install path in CI

duncancmt committed:
Pin foundry-zksync version in CI

duncancmt committed:
Merge branch 'master' into dcmt/upgrade-timelock-1.4.0-multicall

duncancmt committed:
Clean up slop

duncancmt committed:
Fix merge artifact

duncancmt committed:
Omit appended CBOR metadata in ZK mode too

jparklev commented on src/deployer/SafeGuard.sol:
```diff
@@ -0,0 +1,1172 @@
+
+            uint256 threshold = _safe.getThreshold();
+            if (threshold < _MINIMUM_THRESHOLD) {
+                revert ThresholdTooLow(threshold);
```
For a 3 of 5 safe, going through the resignation flow: 

`_getThresholdAfterResign` returns 3 -> `removeOwner` executes -> 3 of 4 -> this check hits (4 - 3 < 2) -> ThresholdTooHigh

Similar for 4 of 6, etc.

Is this intentional? I assume we want to keep at least some buffer of signers in case keys are lost or inaccessible, but this might block `removeOwner`s needed for `cancel` at common multisig configurations?

duncancmt replied:
Let's consider the reverse? What if that scenario were to occur, that the safe had an owner removed and became 3-of-4? How does this affect the recovery game in both the "leaked key" and the "malicious signer" cases?

duncancmt replied:
This is fixed now.

duncancmt committed:
Add support for EIP-7997 factory

Given that EIP-8141 is CFI for Hegota and it depends on 7997, compatibility is desirable.

duncancmt committed:
Add some extra reentrancy checks

duncancmt committed:
Forbid ERC20 gas sponsorship for transactions that mutate the Guard

duncancmt committed:
`forge fmt`

duncancmt committed:
Compilation errors

duncancmt committed:
Fix misleading comment

duncancmt committed:
Require completed Safe{Wallet} deployment in `constructor`

duncancmt committed:
Golf

duncancmt committed:
Add a more robust mechanism for preventing owners from canceling their own ouster

duncancmt committed:
Add a comment explaining why an audit finding won't be fixed

duncancmt committed:
Forbid spamming `enqueue` on transactions with past nonces

duncancmt committed:
Cleanup

duncancmt committed:
Reorganize storage layout

duncancmt committed:
Improve storage locality further

duncancmt committed:
Golf

duncancmt committed:
`enableModule` must also disable sponsorship to avoid module concealment

duncancmt committed:
Also disable sponsorship of `setFallbackHandler`

duncancmt committed:
Comment

duncancmt committed:
Add Bailsec audit for Safe Guard with Multicall support

duncancmt committed:
Decrease threshold to pass later `_MINIMUM_THRESHOLD` checks during forced resignation

duncancmt committed:
Simplify by using inheritance

duncancmt committed:
Golf

duncancmt committed:
Fix CI

duncancmt committed:
Fix CI

duncancmt committed:
Fix CI

duncancmt committed:
`unchecked`

duncancmt committed:
Visibility

jparklev reviewed (approved):
LGTM.

Re "...document that the initial configuration of the Safe should be 2/5 at minimum and in general more owners are better" (from Slack): the Safe (`0xf36b…2F51`) is still 2/4. Are we planning to add a 5th signer before installing the guard?
